src/github.com/matrix-org/dendrite/clientapi/auth/auth.go

```diff
@@ -43,6 +43,27 @@ type DeviceDatabase interface {
 	GetDeviceByAccessToken(ctx context.Context, token string) (*authtypes.Device, error)
 }
 
+// GetUserFromRequest extracts UserID from the given HTTP request,
+// if access token is provided (and correct) or if user_id is given as a parameter of the HTTP request
+// else returns resErr (an error response which can be sent to the client)
+func GetUserFromRequest(req *http.Request, deviceDB DeviceDatabase) (user string, resErr *util.JSONResponse) {
+	dev, err := VerifyAccessToken(req, deviceDB)
+
+	// If access_token is empty try to get UserID from request parameters
+	if err != nil {
```

Comment:
The access token will still be provided by appservices, and should be validated to be from an appservice before considering the value of `user_id`.

Reply:
So appsevices request will be same as https://matrix.org/docs/spec/application_service/unstable.html#identity-assertion ?
Also if appsevices token is invalid, we should reject the whoami ?

Reply:
yes and yes.

For every client/server API endpoint (including whoami) the appservice must specify it's access token included in the registration file. It may then optionally include the user_id to masquerade as that user (assuming it's also in it's namespace). Otherwise it's supposed to use the `sender_localpart` in the registration file as a default.

This endpoint also needs to make sure that the user, whether it be an appservice user or otherwise, is actually registered on the server. For regular users it's not that hard, but for appservice users it can be a bit annoying.

Reply:
@APwhitehat An application service is stored as the following struct: https://github.com/matrix-org/dendrite/blob/6b5597218330b3702cf2b86b2163d5fe769b76a2/src/github.com/matrix-org/dendrite/common/config/appservice.go#L40-L53

Application Services are stored in config.Dendrite.Derived.ApplicationServices: https://github.com/matrix-org/dendrite/blob/6b5597218330b3702cf2b86b2163d5fe769b76a2/src/github.com/matrix-org/dendrite/common/config/config.go#L231-L233

So first iterate through those and check that the token matches a known AppService, then cross reference that with the given user ID in the database and check that it matches the given AppService ID. Users, including AppService users, are stored with the following SQL:

https://github.com/matrix-org/dendrite/blob/6b5597218330b3702cf2b86b2163d5fe769b76a2/src/github.com/matrix-org/dendrite/clientapi/auth/storage/accounts/accounts_table.go#L43-L44

Reject if something doesn't match up.

AppService implementation is still incomplete, but it should be complete enough for this task to work.